ze/btx_zeinterval_callbacks.cpp

```diff
@@ -0,0 +1,727 @@
+    size_t tmp = its->second - its->first;
+    rs.erase(its);
+    return tmp;
+  }
+  auto &rd = state->rangeset_memory_device[hp];
+  auto itd = rd.find(ptr);
+  if (itd != rd.end()) {
+    size_t tmp = itd->second - itd->first;
+    rd.erase(itd);
+    return tmp;
+  }
+  auto &rh = state->rangeset_memory_host[hp];
+  auto ith = rh.find(ptr);
+  if (ith != rh.end()) {
+    size_t tmp = ith->second - ith->first;
+    rh.erase(ith);
+    return tmp;
+  }
+  return 0;
+}
+
+static bool memory_include(const memory_interval_t &m, const uintptr_t val) {
+  // if m empty, begin == end
+  const auto it = m.upper_bound(val);
+  if (it == m.cbegin())
+    return false;
+  return (val < (std::prev(it)->second));
+}
+
+static const char *memory_location(data_t *state, hp_t hp,
+                                   const uintptr_t ptr) {
+  const auto &rs = state->rangeset_memory_shared[hp];
+  if (memory_include(rs, ptr))
+    return "S";
+
+  const auto &rd = state->rangeset_memory_device[hp];
+  if (memory_include(rd, ptr))
+    return "D";
+
+  const auto &rh = state->rangeset_memory_host[hp];
+  if (memory_include(rh, ptr))
+    return "H";
+
+  return "M";
+}
+
+/*
+ * Clock buiness
+ */
+
+static uint64_t convert_device_cycle(
+    uint64_t device_cycle, const clock_lttng_device_t &timestamp_pair_ref,
+    const uint64_t lttng_min, const ze_device_properties_t &device_property) {
+
+  const auto &[lttng_ref, device_cycle_ref] = timestamp_pair_ref;
+
+  assert(device_property.kernelTimestampValidBits <= 64);
+  const uint64_t device_cycle_max_val =
+      ((uint64_t)1 << device_property.kernelTimestampValidBits) - 1;
+
+  uint64_t lttng;
+  const uint64_t device_ref_ns = (device_cycle_ref & device_cycle_max_val) *
+                                 device_property.timerResolution;
+  device_cycle &= device_cycle_max_val;
+  do {
+    const uint64_t device_ns = device_cycle * device_property.timerResolution;
+    lttng = device_ns + (lttng_ref - device_ref_ns);
+    device_cycle += device_cycle_max_val;
+  } while (lttng < lttng_min);
+  return lttng;
+}
+
+static uint64_t
+compute_and_convert_delta(uint64_t start, uint64_t end,
+                          const ze_device_properties_t &device_property) {
+
+  assert(device_property.kernelTimestampValidBits <= 64);
+  const uint64_t max_val =
+      ((uint64_t)1 << device_property.kernelTimestampValidBits) - 1;
+  start &= max_val;
+  end &= max_val;
+  const uint64_t delta =
+      (end >= start) ? (end - start) : (max_val + (end - start));
+  return delta * device_property.timerResolution;
+}
+
+// Hash UUID
+static uintptr_t hash_device(const ze_device_properties_t &device_property) {
+  const auto device_uuid = device_property.uuid;
+
+  // Right now we store `ptr` and not `uuid`.
+  // we cast `uiid` to `ptr`. This should to the opposite
+  uintptr_t device_uuid_hash[2];
+  memcpy(&device_uuid_hash[0], &device_uuid, sizeof(device_uuid));
+  device_uuid_hash[0] += device_uuid_hash[1];
```

Comment:
You copy the value of `device_uuid` into `device_uuid_hash[0]`, but I do not see where `device_uuid_hash[1]` get a valid value.

Reply:
`  memcpy(&device_uuid_hash[0], &device_uuid, sizeof(device_uuid));`  `device_uuid` have the size of `device_uuid_hash+device_uuid_hash` I think (devuce_uuid is a tuple if I remember correcly)